### PR TITLE
[Messenger] fix compatibility with Doctrine DBAL 4

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | 
| License       | MIT

### CODE REVIEW

**Comment by xabbuh** [comment]
see doctrine/dbal#6191

**Comment by derrabus** [comment]
The idea behind that feature was that Symfony could leverage the query builder for building the "skip locked rows" query.

**Comment by xabbuh** [comment]
Thanks for your feedback. I guess it's better now.

File: src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php [hunk 1]
```diff
@@ -177,7 +178,16 @@ public function get(): ?array
...
+
+            if (method_exists(QueryBuilder::class, 'forUpdate')) {
+                $query->forUpdate();
+                try {
+                    $sql = $query->getSQL();
+                } catch (Exception $e) {
+                }
+            }
+
+            if (!method_exists(QueryBuilder::class, 'forUpdate') && preg_match('/FROM (.+) WHERE/', (string) $sql, $matches)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php]:
```suggestion
            } elseif (preg_match('/FROM (.+) WHERE/', (string) $sql, $matches)) {
```

File: src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php [hunk 2]
```diff
@@ -194,8 +204,12 @@ public function get(): ?array
             }
 
             // use SELECT ... FOR UPDATE to lock table
+            if (!method_exists(QueryBuilder::class, 'forUpdate')) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php:207]: We cannot lock anymore in DBAL4?

File: src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php [hunk 3]
```diff
@@ -177,7 +178,16 @@ public function get(): ?array
 
             // Append pessimistic write lock to FROM clause if db platform supports it
             $sql = $query->getSQL();
-            if (preg_match('/FROM (.+) WHERE/', (string) $sql, $matches)) {
+
+            if (method_exists(QueryBuilder::class, 'forUpdate')) {
+                $query->forUpdate();
+                try {
+                    $sql = $query->getSQL();
+                } catch (Exception $e) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php]: can this be a bit more specific? I don't like swallowing any exceptions :grimacing:

Comment (nicolas-grekas) [reply src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php]: !
```suggestion
                } catch (DBALException $e) {
```

File: src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php (hunk 2 above)
Comment (xabbuh) [reply src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php:207]: as far as I understand the related changes in Doctrine DBAL this now happens with the `forUpdate()` call above

File: src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php (hunk 3 above)
Comment (xabbuh) [reply src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php]: https://github.com/doctrine/dbal/pull/6191#discussion_r1379382594

File: src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php (hunk 2 above)
Comment (stof) [reply src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php:207]: indeed

File: src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php [hunk 4]
```diff
@@ -177,7 +177,14 @@ public function get(): ?array
 
             // Append pessimistic write lock to FROM clause if db platform supports it
             $sql = $query->getSQL();
-            if (preg_match('/FROM (.+) WHERE/', (string) $sql, $matches)) {
+
+            if (method_exists(QueryBuilder::class, 'forUpdate')) {
+                $query->forUpdate();
```
Comment (stof) [inline src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php:192]: For the case of the oracle platform, this is now applying the lock on the inner query and not the outer one (see the specific code below). Is this expected ?

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php (hunk 4 above)
Comment (xabbuh) [reply src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php:192]: it's probably better not to change the outcome here, updated

**Review by stof — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @xabbuh.

**Pull request merged**